src/spaniel-instance.ts

```diff
@@ -0,0 +1,199 @@
+  IntersectionObserverInit
+} from './interfaces';
+
+import {
+  EventStore,
+  RAFEventRecord,
+  GenericEventRecord
+} from './metal/events';
+
+import {
+  Scheduler,
+  Frame,
+  generateToken,
+  ElementScheduler
+} from './metal/index';
+
+import {
+  FrameInterface
+} from './metal/interfaces';
+import { Watcher, WatcherConfig } from './watcher';
+
+import w from './metal/window-proxy';
+
+let IntersectionObserver: IntersectionObserverClass = !!w.IntersectionObserver ? w.IntersectionObserver : undefined;
+
+export interface SpanielIntersectionObserverInterface {
+  new(callback: Function, options: IntersectionObserverInit): SpanielIntersectionObserver;
+}
+
+function generateIntersectionObserverClass(parent: SpanielInstance): SpanielIntersectionObserverInterface {
+  return class SpanielIntersectionObserverClass extends SpanielIntersectionObserver {
+    constructor(callback: Function, options: IntersectionObserverInit = {}) {
+      super(callback, options, parent);
+    }
+  };
+}
+
+export interface SpanielObserverInterface {
+  new (callback: (entries: SpanielObserverEntry[]) => void, options: SpanielObserverInit): SpanielObserver;
+}
+
+function generateSpanielObserver(parent: SpanielInstance): SpanielObserverInterface {
+  return class SpanielObserverClass extends SpanielObserver {
+    constructor(callback: (entries: SpanielObserverEntry[]) => void, options: SpanielObserverInit = {}) {
+      super(callback, options, parent);
+    }
+  };
+}
+
+export interface WatcherInterface {
+  new(config: WatcherConfig): Watcher;
+}
+
+function generateWatcher(parent: SpanielInstance): WatcherInterface {
+  return class WatcherClass extends Watcher {
+    constructor(config: WatcherConfig = {}) {
+      super(config, parent);
+    }
+  };
+}
+
+export class SpanielInstance {
+  public IntersectionObserver: IntersectionObserverClass;
+  private engine: Engine;
+  private scheduler: Scheduler;
+  public scheduleRead: (callback: Function) => void;
+  public scheduleWork: (callback: Function) => void;
+  public eventStore: EventStore;
+  public SpanielObserver: SpanielObserverInterface;
+  public Watcher: WatcherInterface;
+  constructor(UserIntersectionObserver?: IntersectionObserverClass, engine?: OptimizedEngineInterface ) {
+    if (engine) {
+      this.engine = new OptimizedEngine(engine.onMutate, engine.scheduleWork);
+    } else {
+      this.engine = new Engine();
+    }
+    this.scheduler = new Scheduler(this.engine);
+    this.scheduleRead = this.scheduler.scheduleRead;
+    this.scheduleWork = this.scheduler.scheduleWork;
+    this.IntersectionObserver = UserIntersectionObserver || (IntersectionObserver) ? IntersectionObserver : generateIntersectionObserverClass(this);
+    this.SpanielObserver = generateSpanielObserver(this);
+    this.Watcher = generateWatcher(this);
+    if (w.hasDOM) {
+      window.addEventListener('unload', (e: any) => {
+        // First fire internal event to fire any observer callbacks
```

Comment:
who cleans up this global callbacks? `unload` and `visibilitychange`, does the `SpanielInstance` (hopefully named Session` need a `destroy` for cleanup?